File: lab4-records/java/Drinks.java
public class Drinks {

  // A Java union type
  public static abstract class DrinkCategory{
    public abstract String drinkColour();
  }

  public static class Water extends DrinkCategory {
    public String drinkColour(){ return "clear"; }
  }

  public static class Milk extends DrinkCategory {
    public double fatContent;
    public Milk(double fatContent){ this.fatContent = fatContent; }
    public String drinkColour(){ return "white"; }
  }
  public static class Juice extends DrinkCategory {
    public String fruit;
    public Juice(String fruit){ this.fruit = fruit; }
    public String drinkColour(){ 
      switch(fruit){
        case "orange": return "orange";
        case "lemon": return "yellow";
        default: return "unknown";
      }
    }
  }

  public DrinkCategory drink1 = new Water();
  public DrinkCategory drink2 = new Milk(3);
  public DrinkCategory drink3 = new Milk(1);
  public DrinkCategory drink4 = new Juice("orange");
}

File: lab4-records/haskell/Drinks.hs
module Drinks where

data DrinkCategory = Water | Milk { fatContent :: Double } | Juice { fruit :: String }

drink1 = Water
drink2 = Milk 3
drink3 = Milk 1
drink4 = Juice "orange"

drinkColour :: DrinkCategory -> String
drinkColour Water = "clear"
drinkColour (Milk _) = "white"
drinkColour (Juice "orange") = "orange"
drinkColour (Juice "lemon") = "yellow"
drinkColour _ = "unknown"

